---
title: "EnImpute2_bib"
author: "Yu"
date: "2023-01-17"
output:
  pdf_document: default
  html_document: default
bibliography: EnImpute2.bib
csl: "bmc-bioinformatics.csl"
nocite: |
  @Linderman397588,@eraslan2018single, @kwak2018drimpute,@wagner2018k,@Dijk2018Recovering,@Huang2018SAVER,@Li2018an,@ye2019scnpf,@peng2019scrabble,@10.1093/bioinformatics/btaa139,@10.1093/bioinformatics/btaa108,@li2022sctssr2,@chen2018viper,@10.1371/journal.pcbi.1009118,@Risso125112,@10.1093/bioinformatics/btz435,@wang2018bayesian,@Satija2015Spatial,@Chang2017Web,
---

## Reference

1.  Linderman GC et al. Zero-preserving imputation of scRNA-seq data using low-rank approximation. bioRxiv. 2018. <https://doi.org/10.1101/397588>.
2.  Eraslan G et al. Single cell RNA-seq denoising using a deep count autoencoder. bioRxiv. 2018;300681.
3.  Kwak I-Y et al. DrImpute: Imputing dropout events in single cell RNA sequencing data. BMC Bioinformatics. 2018;19:220.
4.  Wagner F et al. K-nearest neighbor smoothing for high-throughput single-cell RNA-seq data. bioRxiv. 2018;217737.
5.  Dijk D van et al. Recovering gene interactions from single-cell data using data diffusion. Cell. 2018;174:1--4.
6.  Huang M et al. SAVER: Gene expression recovery for single-cell RNA sequencing. Nat Methods. 2018;15:539--42.
7.  Li WV, Li JJ. An accurate and robust imputation method scImpute for single-cell RNA-seq data. Nat Commun. 2018;9:997.
8.  Ye W, Ji G, Ye P, Long Y, Xiao X, Li S, et al. scNPF: An integrative framework assisted by network propagation and network fusion for preprocessing of single-cell RNA-seq data. Bmc Genomics. 2019;20:1--6.
9.  Peng T, Zhu Q, Yin P, Tan K. SCRABBLE: Single-cell RNA-seq imputation constrained by bulk RNA-seq data. Genome biology. 2019;20:1--2.
10. Chen C, Wu C, Wu L, Wang X, Deng M, Xi R. scRMD: imputation for single cell RNA-seq data via robust matrix decomposition. Bioinformatics. 2020;36:3156--61.
11. Jin K, Ou-Yang L, Zhao X-M, Yan H, Zhang X-F. scTSSR: gene expression recovery for single-cell RNA sequencing using two-side sparse self-representation. Bioinformatics. 2020;36:3131--8.
12. Li B, Jin K, Ou-Yang L, Yan H, Zhang X-F. scTSSR2: Imputing dropout events for single-cell RNA sequencing using fast two-side self-representation. IEEE/ACM Transactions on Computational Biology and Bioinformatics. 2022.
13. Chen M, Zhou X. VIPER: Variability-preserving imputation for accurate gene expression recovery in single-cell RNA sequencing studies. Genome biology. 2018;19:1--5.
14. Qi YAZ Jing AND Zhou. SDImpute: A statistical block imputation method based on cell-level and gene-level information for dropouts in single-cell RNA-seq data. PLOS Computational Biology. 2021;17:1--20.
15. Risso D, Perraudeau F, Gribkova S, Dudoit S, Vert J-P. ZINB-WaVE: A general and flexible method for signal extraction from single-cell RNA-seq data. bioRxiv. 2017. <https://doi.org/10.1101/125112>.
16. Zhang X-F, Ou-Yang L, Yang S, Zhao X-M, Hu X, Yan H. EnImpute: imputing dropout events in single-cell RNA-sequencing data via ensemble learning. Bioinformatics. 2019;35:4827--9.
17. Wang X, Erdman C, Emerson J. Bayesian analysis of change point problems. R-CRAN Package, r-project org Available at: <https://cran> r-project org/web/packages/bcp/bcp pdf. 2018.
18. Satija R et al. Spatial reconstruction of single-cell gene expression data. Nat Biotechnol. 2015;33:495--502.
19. Chang W et al. Web application framework for r. Spell Press; 2017.
